### PR TITLE
feat: Add support for attached properties localization

### PR DESCRIPTION
**Pull request by jeromelaban**
GitHub Issue (If applicable): closes https://github.com/unoplatform/uno/issues/921, closes https://github.com/unoplatform/uno/issues/9426

## PR Type

What kind of change does this PR introduce?
- Feature

## What is the new behavior?

Add support for attached properties localization.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs [hunk 1]
```diff
@@ -651,6 +652,22 @@ private string[] GetResourceKeys(CancellationToken ct)
 			return resourceKeys;
 		}
 
+		private string RewriteResourceKeyName(StringBuilder builder, string keyName)
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs:655]: Consider making `static` if it doesn't reference instance members. [CA1822](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1822)

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs [hunk 2]
```diff
@@ -651,6 +652,22 @@ private string[] GetResourceKeys(CancellationToken ct)
...
+		{
+			var firstDotIndex = keyName.IndexOf('.');
+			if (firstDotIndex != -1)
+			{
+				builder.Clear();
+				builder.Append(keyName);
+
+				builder[firstDotIndex] = '/';
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs:663]: Previously we were replacing all `.`s with `/`s
Now we replace only the first `.`, not sure if that's intentional.

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.Reflection.cs [hunk 3]
```diff
@@ -899,6 +899,44 @@ private IEnumerable<string> FindLocalizableProperties(XamlType xamlType)
...
+				if(propertyPath.StartsWith(usingPattern))
+				{
+					var lastDotIndex = propertyPath.LastIndexOf('.');
+
+					var propertyName = propertyPath.Substring(lastDotIndex + 1);
+					var typeName = propertyPath
+						.Substring(usingPattern.Length, lastDotIndex - usingPattern.Length)
+						.Replace("]", ".");
+
+					if(GetType(typeName) is { } typeSymbol)
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.Reflection.cs:929]: It looks like `typeName` here is fully qualified name? In that case, `GetType` is much less performant than `_metadataHelper.GetTypeByFullName`

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs [hunk 4]
```diff
@@ -4888,13 +4912,12 @@ string Inner()
...
-			if (owner != null && IsAttachedProperty(owner))
+			if (owner != null && IsAttachedProperty(owner, memberName))
 			{
-				var declaringType = GetType(owner.Member.DeclaringType);
+				var declaringType = owner;
 				var nsRaw = declaringType.ContainingNamespace.GetFullName();
-				var ns = nsRaw?.Replace(".", "/");
 				var type = declaringType.Name;
-				fullKey = $"{uidName}/[using:{ns}]{type}/{memberName}";
+				fullKey = $"{uidName}/[using:{nsRaw}]{type}.{memberName}";
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:4920]: Do we have any tests for when `ContainingNamespace` is the global namespace?

File: src/Uno.UWP/ApplicationModel/Resources/ResourceLoader.cs [hunk 5]
```diff
@@ -278,11 +286,29 @@ private static void ProcessResourceFile(string fileName, Stream input, string[]
...
+					StringBuilder sb = new();
 					for (var i = 0; i < resourceCount; i++)
 					{
 						var key = reader.ReadString();
 						var value = reader.ReadString();
 
+						if (adjustKeyTransformationForV2)
+						{
+							// Restore the original format
+							key = key.Replace("/", ".");
```
Comment (Youssef1313) [inline src/Uno.UWP/ApplicationModel/Resources/ResourceLoader.cs:298]:
```suggestion
							key = key.Replace('/', '.');
```

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs (hunk 2 above)
Comment (jeromelaban) [reply src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs:663]: It's intentional, yes. That's the way resources are looked up using ResourceLoader.

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs (hunk 4 above)
Comment (jeromelaban) [reply src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:4920]: In that particular case, the global namespace cannot be used (it's supposed to come from a type that can only be defined from a non-empty namespace).

**Review by dr1rrb — APPROVED** [review]

File: src/Uno.UWP/ApplicationModel/Resources/ResourceLoader.cs [hunk 6]
```diff
@@ -259,7 +260,14 @@ private static void ProcessResourceFile(string fileName, Stream input, string[]
...
 				var version = reader.ReadInt32();
+				if(version == 2)
+				{
+					version = UPRIVersion;
+					adjustKeyTransformationForV2 = true;
+				}
+
 				if (version != UPRIVersion)
 				{
 					throw new InvalidOperationException($"The resource file {fileName} has an invalid version (got {version}, expecting {UPRIVersion})");
```
Comment (carldebilly) [inline src/Uno.UWP/ApplicationModel/Resources/ResourceLoader.cs:273]: Why adding code to support v2 if you're immediately throwing an unsupported version right after?

**Review by carldebilly — APPROVED** [review]

**Pull request merged**